presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -326,7 +331,73 @@ public static boolean canCoerce(Type actualType, Type expectedType)
     public static boolean canCoerce(TypeSignature actualType, TypeSignature expectedType)
     {
         Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperTypeSignature(actualType, expectedType);
-        return commonSuperTypeSignature.isPresent() && commonSuperTypeSignature.get().equals(expectedType);
+        return commonSuperTypeSignature.isPresent() && typesMatch(commonSuperTypeSignature.get(), expectedType);
+    }
+
+    /**
+     * Checks type signatures for equality with the only difference:
+     * <p>
+     * If the long parameters are not specified in `expectedType` than any calculated long values will be accepted
+     * <p>
+     * e.g.
+     * <p>
+     * typesMatch(DECIMAL(2,1), DECIMAL) -> TRUE
```

Comment:
Actually it is more about specified decimal and unspecified. We can declare function signature with specified decimal, like `DECIMAL(2,1)`, and with unspecified, like any `DECIMAL(x,y)`. So, `typesMatch(DECIMAL(x,y), DECIMAL(2,1))->TRUE`, `typesMatch(DECIMAL(3,2), DECIMAL(2,1))->FALSE`